### ISSUE TITLE
Potential inconsistency between `_mm_loadl_epi64` and `_mm_loadu_si64`

### ISSUE THREAD

**Issue by marmeladema** [rust-lang/stdarch#1166]
## Description

According to Intel documentation at:

* https://software.intel.com/sites/landingpage/IntrinsicsGuide/#text=_mm_loadu_si64&expand=3421

and

* https://software.intel.com/sites/landingpage/IntrinsicsGuide/#text=_mm_loadl_epi64&expand=3357

Both intrinsics should be equivalent and load a 64-bit integer from memory into the first element of dst.
They also both have the same pseudo code:
```
dst[63:0] := MEM[mem_addr+63:mem_addr]
dst[MAX:64] := 0
```
They should also be both be implemented with the same instruction:
```
movq xmm, m64
```
The differences are that the first is SSE and accept a `void const* mem_addr` as argument whereas the second is SSE2 and accept a `__m128i const* mem_addr` as argument.

## Expected behavior:

I might be wrong, but I would expect both intrinsics to do exactly the same thing.

## Observed behavior:

[I tried this little code snippet in the playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=29a5787481863591ce4457c5eb650e9b):
```Rust
#[cfg(target_arch = "x86_64")]
use std::arch::x86_64::*;

fn print_m128i(vector: __m128i) {
    println!("{:?}", unsafe { std::mem::transmute::<__m128i, [i8; 16]>(vector) })
}

fn main() {
    let v1 = unsafe { _mm_set1_epi8(127) };
    print_m128i(v1);
    // Additional 9th byte at the end as a sentinel value. It should be ignored by both intrinsics.
    let arr = [8i8, 7i8, 6i8, 5i8, 4i8, 3i8, 2i8, 1i8, 9i8];
    let v3 = unsafe { _mm_loadu_si64(&arr as *const i8 as *const u8) };
    print_m128i(v3);
    let v2 = unsafe { _mm_loadl_epi64(&arr as *const i8 as *const __m128i) };
    print_m128i(v2);
}
```

and for `_mm_loadu_si64` it prints:

```
[0, 0, 0, 0, 0, 0, 0, 0, 8, 7, 6, 5, 4, 3, 2, 1]
```

whereas for `_mm_loadl_epi64` it prints:
```
[8, 7, 6, 5, 4, 3, 2, 1, 0, 0, 0, 0, 0, 0, 0, 0]
```

I *think* (but please correct me if I am wrong) that the correct implementation is `_mm_loadl_epi64`.

Have I misinterpreted something?
I was *very confused* when trying those intrinsics.
If this is bug, is it fixable?
If this is expected behavior maybe the documentation could be enhanced somehow?

**Comment by Lokathor** [comment]
Given that your array is not necessarily aligned, and that loadu allows for unaligned reads while loadl does not, I'm at least a little suspicious of the test without knowing if that array if i8 happened to end up being aligned to 8 bytes or not (which it might do by chance).

**Comment by marmeladema** [comment]
About the unaligned load, I have read in multiple places that `_mm_loadl_epi64` allows unaligned load, even if not explicitly stated by Intel documentation:
* https://fgiesen.wordpress.com/2016/04/03/sse-mind-the-gap

Additionally, stdarch itself seems to believe this too https://github.com/rust-lang/stdarch/pull/584

**Comment by marmeladema** [comment]
I also have found this about the unaligned-ness of `_mm_loadl_epi64`: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=70118

**Comment by marmeladema** [comment]
What about [this](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=3c6929684ca32948f7689167547f3340):
```Rust
#[cfg(target_arch = "x86_64")]
use std::arch::x86_64::*;

fn print_m128i(vector: __m128i) {
    println!("{:?}", unsafe { std::mem::transmute::<__m128i, [i8; 16]>(vector) })
}

#[repr(align(16))]
struct Array([i8; 8]);

fn main() {
    let arr = Array([8i8, 7i8, 6i8, 5i8, 4i8, 3i8, 2i8, 1i8]);
    let arr_ptr = &arr.0 as *const i8;
    println!("arr_ptr: {:?}", arr_ptr);
    let v3 = unsafe { _mm_loadu_si64(arr_ptr as *const u8) };
    println!("_mm_loadu_si64:");
    print_m128i(v3);
    let v2 = unsafe { _mm_loadl_epi64(arr_ptr as *const __m128i) };
    println!("_mm_loadl_epi64:");
    print_m128i(v2);
}
```

It prints:
```
arr_ptr: 0x7ffc63873110
_mm_loadu_si64:
[0, 0, 0, 0, 0, 0, 0, 0, 8, 7, 6, 5, 4, 3, 2, 1]
_mm_loadl_epi64:
[8, 7, 6, 5, 4, 3, 2, 1, 0, 0, 0, 0, 0, 0, 0, 0]
```

And if I am not mistaken, `0x7ffc63873110` is 16-bytes aligned right?

**Comment by Amanieu** [comment]
`_mm_loadu_si64` was fixed in #992, but this hasn't made its way to stable yet.

**Comment by marmeladema** [comment]
Oh! Thanks for letting me know! I did rerun the code snippets with nightly and it seems fixed indeed!
